### DISCUSSION TITLE
Handling a control change (all note off) when using your filtering technique

### DISCUSSION THREAD

**Discussion by fapplin** [FortySevenEffects/arduino_midi_library#342] [Q&A]
Hi Francois,

I hope this is the last question for a while.  I am using your setThruFilter and setThruMap methods. I am redefining how my white keys and black keys work on the fly. You helped with the outline of the following code (see below).  I handle all my NoteOn and NoteOff (I only use white keys for playing) and I can use my black keys to do simple program changes (change instruments). This all works great (thank you). But, I have one other need and that is to be able to handle just one or two case of control changes - all notes off and all sounds off. I have a Panic button on my user interface and when they press it, I want to do the sendControlChange() myself. How do I add that functionality to  setThruFilter() and setThruMap() methods?

Here's what's working.

```c++
bool forwardToMidiUSB(const midi::Message<128>& message)
{ 
  //TRACE_ME_IN();
  
  const bool forward = message.type != midi::NoteOn && message.type != midi::NoteOff;
  
  // Forward only NoteOn for C#3, no NoteOff, and every other type of message
  bool forwardToThru = message.type != midi::NoteOff;
  if (message.type == midi::NoteOn && message.data1 != kmidi.getBlackKeysProgramChange() )
  {
    forwardToThru = false;
  }
  if (forward)
  {
    //Serial.printf("forwarding message\n");
    midiUSB.send(message);
  }
  // Thru back to the Serial output

   //TRACE_ME_OUT();
  return forwardToThru;
}

midi::Message<128> mapNoteOnToPC(const midi::Message<128>& message)
{
  //TRACE_ME_IN();
  
  midi::Message<128> out(message);
  if (message.type == midi::NoteOn && message.data1 == kmidi.getBlackKeysProgramChange() /* C#4 */)
  {
    out.type = midi::ProgramChange;
    out.data1 = kmidi.getBlackKeysProgramChange(); // Or whatever
  } //endif

   //TRACE_ME_OUT();
  return out;
}
```

**Comment by franky47** [comment]
I don't think you need to add it to those methods, if you have a physical button connected to the Arduino, it's a matter of:
1. Reading the digital input it's attached to
2. Detecting state changes (watch out for debouncing issues, there are a few libraries that provide debouncing digital inputs out there).
3. When a pressed change is detected, call `MIDI.sendControlChange` with the appropriate parameters.

Filter and map are only here to alter the flow of incoming MIDI messages through the Arduino, we're kind of hacking them to fork to MIDI USB but ideally they should be free of side-effects.

**Reply by fapplin** [reply]
I should've added that the screen doing the Panic button is not connected to the Pico I'm using to do the MIDI processing. I send commands from a RPi, which is hitched up to a touch display. The RPi sends commands via serial to the Pico. One core is handling serial input from the RPi and one core is doing the MIDI processing. So, if I send the command to turn "all notes off" from my controller - it works fine. If I send the command from my touch screen - I see it getting called on the Pico correctly - this is where I do sendControlChange(), but the "stuck note" doesn't get turned off. That's why I suspected it was just getting pushed through the filtering process.

**Reply by fapplin** [reply]
OK. Maybe I'm not formatting the sendControlChange() correctly. This is what I have.

```
midiUSB.sendControlChange(7, 123,  midiUSB.getChannel());
midiUSB.sendControlChange(7, 0,  midiUSB.getChannel());
```

Your docs say: 

```
void MidiInterface< Transport, Settings, Platform >::sendControlChange 	( 	DataByte  	inControlNumber,
		DataByte  	inControlValue,
		Channel  	inChannel 
	) 	
```

When I send the "all notes off" from the controller, this is what I see via

```
Serial.printf("message.type: %d, message.data1: %d, message.data2: %d\n", message.type, message.data1, message.data2);
```

in the bool forwardToMidiUSB(const midi::Message<128>& message).

```
message.type: 176, message.data1: 7, message.data2: 123
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 254, message.data1: 0, message.data2: 0
message.type: 176, message.data1: 7, message.data2: 0
```
Am I doing the control change correctly?

**Reply by fapplin** [reply]
OK. I have it. My code was correct above - the channel wasn't getting set properly (my fault).
